main.nf

```diff
@@ -0,0 +1,533 @@
+log.info "========================================="
+
+
+def create_workflow_summary(summary) {
+
+    def yaml_file = workDir.resolve('workflow_summary_mqc.yaml')
+    yaml_file.text  = """
+    id: 'nf-core-deepvariant-summary'
+    description: " - this information is collected when the pipeline is started."
+    section_name: 'nf-core/deepvariant Workflow Summary'
+    section_href: 'https://github.com/nf-core/deepvariant'
+    plot_type: 'html'
+    data: |
+        <dl class=\"dl-horizontal\">
+${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style=\"color:#999999;\">N/A</a>'}</samp></dd>" }.join("\n")}
+        </dl>
+    """.stripIndent()
+
+   return yaml_file
+}
+
+
+/********************************************************************
+  process preprocessFASTA
+  Collects all the files related to the reference genome, like
+  .fai,.gz ...
+  If the user gives them as an input, they are used
+  If not they are produced in this process given only the fasta file.
+********************************************************************/
+
+if(!fai) {
+  process preprocessFAI {
+      publishDir "$baseDir/sampleDerivatives"
+
+      input:
+      file(fasta) from fastaCh1
+
+      output:
+      file("${fasta}.fai") into faiCh
+
+      script:
+      """
+      samtools faidx $fasta
+      """
+  }
+}
+
+if(!fastagz) {
+  process preprocessFASTGZ {
+      publishDir "$baseDir/sampleDerivatives"
+
+      input:
+      file(fasta) from fastaCh2
+
+      output:
+      file("*.gz") into (tmpFastaGzCh, fastaGzCh)
+
+      script:
+      """
+      bgzip -c ${fasta} > ${fasta}.gz
+      """
+  }
+}
+
+if(!gzfai) {
+  process preprocessGZFAI {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaCh3
+    file(fastagz) from tmpFastaGzCh
+
+    output:
+    file("*.gz.fai") into gzFaiCh
+
+    script:
+    """
+    samtools faidx $fastagz
+    """
+  }
+}
+
+if(!gzi){
+  process preprocessGZI {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaCh4
+
+    output:
+    file("*.gz.gzi") into gziCh
+
+    script:
+    """
+    bgzip -c -i ${fasta} > ${fasta}.gz
+    """
+  }
+}
+
+//channel to collect all necessary reference files
+fastaChannel = Channel.from(fastaCh).mix(bedCh, faiCh, fastaGzCh, gzFaiCh, gziCh).collect()
+
+/********************************************************************
+  process preprocessBAM
+  Takes care of the read group line.
+********************************************************************/
+
+process preprocessBAM{
+
+  tag "${bam[0]}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  set val(prefix), file(bam) from bamChannel
+
+  output:
+  set file("ready/${bam[0]}"), file("ready/${bam[0]}.bai") into completeChannel
+
+  script:
+  """
+  mkdir ready
+  [[ `samtools view -H ${bam[0]} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { picard AddOrReplaceReadGroups \
+  I=${bam[0]} \
```

Comment:
Not understanding why `bam` is accessed as an array, it seems to a sole file path therefore this should not work.

Reply:
then, why there's `prefix` if it's never used in the command task ?

Reply:
Hi @pditommaso, thanks for your input,

`bam` is accessed as an array to keep it constisent with `bam_folder` which also sets `bamChannel`

```nextflow
if(params.bam_folder) {
  Channel
      .fromPath("${params.bam_folder}/${params.bam_file_prefix}*.bam")
      .ifEmpty { exit 1, "${params.bam_folder}/${params.bam_file_prefix}*.bam not found"}
      .map{ file -> tuple(file.name, file) }
      .set{bamChannel}
} else if(params.bam) {
  Channel
      .fromPath(params.bam)
      .ifEmpty { exit 1, "${params.bam} not found"}
      .map{ file -> tuple(file.name, file) }
      .set{bamChannel}
}
```

I'm sure this isn't the best way of doing it, however, having tested it it does work.

Is there another way that you think might be better?